File: 2787-ways-to-express-an-integer-as-sum-of-powers/2787-ways-to-express-an-integer-as-sum-of-powers.cpp
class Solution {
    int mod = 1e9 + 7;
    
private:
    int countNumWays(int num, int n, int x, vector<vector<int>>& dp) {
        if(n < 0) return 0;
        if(n == 0) return 1;
        if(num > n) return 0;
        
        if(dp[num][n] != -1) return dp[num][n];
    
        int not_take = countNumWays(num + 1, n, x, dp);
        int take = 0;
        if(n - pow(num, x) >= 0) {
            take = countNumWays(num + 1, n - pow(num, x), x, dp);
        }
        
        return dp[num][n] = (take + not_take) % mod;
    }
    
public:
    int numberOfWays(int n, int x) {
        vector<vector<int>> dp (n + 1, vector<int>(n + 1, -1));
        return countNumWays(1, n, x, dp);
    }
};

File: 2787-ways-to-express-an-integer-as-sum-of-powers/2787-ways-to-express-an-integer-as-sum-of-powers.java
class Solution {
    int mod = (int)1e9 + 7;
    
    private int countNumWays(int num, int curr_sum, int n, int x, int dp[][]) {
        if(curr_sum > n) return 0;
        if(curr_sum == n) return 1;
        if(num > n) return 0;
        
        if(dp[num][curr_sum] != -1) return dp[num][curr_sum];
    
        int not_take = countNumWays(num + 1, curr_sum, n, x, dp);
        int take = 0;
        if(curr_sum + Math.pow(num, x) <= n) {
            take = countNumWays(num + 1, curr_sum + (int)Math.pow(num, x), n, x, dp);
        }
        
        return dp[num][curr_sum] = (take + not_take) % mod;
    }
    
    public int numberOfWays(int n, int x) {
        int dp[][] = new int[n + 1][n + 1];
        
        for(int i = 0; i <= n; i++) {
            for(int j = 0; j <= n; j++)
                dp[i][j] = -1;
        }
        
        return countNumWays(1, 0, n, x, dp);
    }
}
    